### PR TITLE
Immutable Record

### PR DESCRIPTION
**Pull request by haynesjm42**
Record immutability is now enforced by freezing the Record object, and deep freezing it's data, immediately after construction.

Also added the ability to update Record fields (instead of just raw data) via Store methods, and tracking of when a Record is "dirty". Per discussion with @lbwexler - updating Record fields puts the Record into a dirty state. Whenever the Record raw data is updated, we assume it is the pristine version of the Record, and the Record will no longer be dirty.

To accommodate inline grid editing now that Records are immutable I've added `editable` and `updateValueFn` as Column configs. We install a default `updateValueFn` which will automatically call `Store.updateRecordFields()` with the new value after inline editing (via ag-Grid `valueSetter` callback)

As part of these changes it might make sense to add inline editing support to RestGrid. I've added basic support to the RestGridPanel in Toolbox for testing/demonstration purposes, but could pretty easily see how this could fit well into RestGrid as an alternative to the form-based editing. Would probably want to implement our own text/number/date/boolean cell editors if we do that.

https://github.com/xh/toolbox/tree/recordUpdate

----------------------------

Hoist P/R Checklist
-------------------

Review and check off the below. Items that do not apply can also be checked off to indicate they
have been considered. If unclear if a step is relevant, please leave unchecked and note in comments.

- [ ] Up to date with `develop` branch as of last change.
- [ ] Ready for review (or open as a draft PR / add `wip` label).
- [ ] Added CHANGELOG entry (or N/A)
- [ ] Reviewed for breaking changes (add `breaking-change` label + CHANGELOG if so, or N/A)
- [ ] Updated doc comments / prop-types (or N/A)
- [ ] Reviewed and tested on Mobile (or N/A)
- [ ] Created Toolbox branch / PR (link provided here, or N/A)

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

File: cmp/grid/columns/Column.js [hunk 1]
```diff
@@ -346,6 +364,12 @@ export class Column {
         const sortCfg = find(this.gridModel.sortBy, {colId: this.colId});
         return sortCfg ? sortCfg.comparator(v1, v2) : agUtils.defaultComparator(v1, v2);
     };
+
+    defaultUpdateFieldFn = ({value, record, store, field}) => {
+        const data = {};
```
Comment (lbwexler) [inline cmp/grid/columns/Column.js]: minor point, but can be done as a one-liner with `const data = {(field):value}`

File: data/Record.js [hunk 2]
```diff
@@ -20,29 +21,44 @@ export class Record {
     id;
     /** @member {(string|number)} */
     parentId;
-    /** @member {Store} */
-    store;
+    /** @member {Record} */
+    originalRecord;
```
Comment (lbwexler) [inline data/Record.js]: might default original to point to `this`, given its name, and then use a === check for isDirty().

File: data/Record.js [hunk 3]
```diff
@@ -114,24 +130,31 @@ export class Record {
...
+
+        forEach(data, (v, key) => {
+            Object.defineProperty(this, key, {
+                get: () => this.data[key],
+                set: () => {
+                    throw XH.exception(`Cannot set read-only field '${key}' on immutable Record. Use Store.updateData() or Store.updateFields().`);
+                }
+            });
         });
+        deepFreeze(this.data);
```
Comment (lbwexler) [inline data/Record.js]: Why the deepFreeze, and the special property definition?

File: cmp/grid/GridModel.js [hunk 4]
```diff
@@ -400,8 +400,8 @@ export class GridModel {
     }
 
     /** Update the underlying store. */
```
Comment (lbwexler) [inline cmp/grid/GridModel.js]: could consider whether to trampoline the additional new store methods in this way as well

File: cmp/grid/columns/Column.js [hunk 5]
```diff
@@ -126,6 +126,8 @@ export class Column {
         exportWidth,
         excludeFromExport,
         tooltip,
+        editable,
```
Comment (lbwexler) [inline cmp/grid/columns/Column.js]: needs some docs here to help evaluated these new args

File: data/Record.js [hunk 6]
```diff
@@ -20,29 +22,57 @@ export class Record {
...
+    get isDirty() {
+        return this.originalRecord !== this;
+    }
+
+    /** @returns {boolean} */
+    get isModified() {
+        return !this.isNew && this.isDirty;
+    }
+
+    /** @returns {Object|null} */
```
Comment (lbwexler) [inline data/Record.js]: perhaps make this a method, since it is doing a fair bit of work and allocation?

File: data/Record.js [hunk 7]
```diff
@@ -114,24 +144,29 @@ export class Record {
...
-            );
-            this[name] = f.parseVal(data[name]);
+        this.originalRecord = withDefault(originalRecord, this);
+        this.raw = raw;
+        this.data = data;
+        this.store = store;
+        this.xhTreePath = this.parent ? [...this.parent.xhTreePath, id] : [id];
+        this.xhIsSummary = isSummary;
+
+        forEach(data, (v, key) => {
```
Comment (lbwexler) [inline data/Record.js]: this suggests a new syntax for working with records, no?

File: data/Record.js [hunk 8]
```diff
@@ -146,10 +181,17 @@ export class Record {
...
             this.store == rec.store &&
-            this.store.fields.every(f => f.isEqual(this[f.name], rec[f.name]))
+            equal(this.data, rec.data)
         );
     }
 
+    isFieldDirty(name) {
+        if (!this.isDirty) return false;
+
+        const value = this.data[name], originalValue = this.originalRecord.data[name];
```
Comment (lbwexler) [inline data/Record.js]: could we use getters on record here?  also, why are removing the field.isEqual check?

File: data/Store.js [hunk 9]
```diff
@@ -138,25 +150,26 @@ export class Store {
...
-                summaryUpdateRec = this.createRecord(rawSummaryData);
+                summaryUpdateRec = this.createRecord({...this.summaryRecord.raw, ...rawSummaryData}, null, true);
             }
         }
 
         // 3) Apply changes
         let didUpdate = false;
         if (!isEmpty(updateRecs) || (addRecs && addRecs.size) || !isEmpty(remove)) {
-            this._all = this._all.updateData({update: updateRecs, add: addRecs, remove: remove});
+            this._original = this._original.loadRecordTransaction({update: updateRecs, add: addRecs, remove: remove});
```
Comment (lbwexler) [inline data/Store.js]: if original and all were the same when this method was called, aren't we doing double the work, and saving an extra representation?

Comment (lbwexler) [reply data/Store.js]: We could consider not allowing the loading of a transaction for a dirty recordset.

**Comment by haynesjm42** [comment]
Opening new PR for latest changes

**Pull request closed without merging**